### ISSUE TITLE
Add a way to determine the index of an n-ary sum

### ISSUE THREAD

**Issue by kosmikus** [well-typed/generics-sop#22]
I regularly find myself writing a function like this:
```
index_NS :: forall f xs . NS f xs -> Int
index_NS = go 0
  where
    go :: forall ys . Int -> NS f ys -> Int
    go !acc (Z _) = acc
    go !acc (S x) = go (acc + 1) x
```

I think it would be useful to add this function to the library.

An additional argument is that if we ever switch to compact internal representations, then `index_NS` is trivial to define. The compact representation of an `NS` is defined as follows:
```
data NS (f :: k -> *) (xs :: [k]) = NS !Int (f Any)
```
So `index_NS` becomes just extracting the first component of the pair. On the other hand, a manual reimplementation of `index_NS` in user code like the one shown above becomes unnecessarily inefficient.

A different option might be to change `collapse_NS` and `collapse_SOP` instead to return the index as well as the stored component. The argument being that while `collapse_NP` and `collapse_POP` are lossless operations, the current definitions of `collapse_NS` and `collapse_SOP` are not. The code would look as follows:
```
collapse_NS  :: forall a xs  .               NS  (K a) xs  ->  (Int, a)
collapse_NS = go 0
  where
    go :: forall ys . Int -> NS (K a) ys -> (Int, a)
    go !acc (Z (K x)) = (acc, x)
    go !acc (S xs)    = go (acc + 1) xs

collapse_SOP :: forall a xss . SListI xss => SOP (K a) xss ->  (Int, [a])
collapse_SOP = collapse_NS . hliftA (K . collapse_NP) . unSOP

type instance CollapseTo NS  a = (Int, a)
type instance CollapseTo SOP a = (Int, [a])

instance HCollapse NS  where hcollapse = collapse_NS
instance HCollapse SOP where hcollapse = collapse_SOP
```

Then `index_NS` can (at least theoretically) be defined in terms of `collapse_NS`:
```
index_NS :: SListI xs => NS f xs -> Int
index_NS = fst . collapse_NS . map_NS (const (K ()))
```

An argument in favour of the adapted `hcollapse` definitions is that with the current representation, if you need to call `hcollapse` anyway, then calling `index_NS` will lead to an additional and unnecessary traversal of the constructors of `NS`. On the other hand, this would be a significant breaking change.

Feedback welcome.

/cc @edsko @phadej

**Comment by phadej** [comment]
I have myself defined
```hs
nsToInt :: NS f xs -> Int
```
so indeed, having it in `generics-sop` is good. `index_NS` is also a good name.

About `hcollapse`: I don't know. I'm slightly :-1: of changing the type I'd rather introduce new name. `hcollapseWithIndex` sounds wrong though.

**Comment by kosmikus** [comment]
@phadej Ok, I think I'll add `index_NS` then, for now. That's less invasive and definitely seems useful.

**Comment by kosmikus** [comment]
Still a number of minor points coming up as a result of thinking about this:

1. Do we want `index_NS` and `index_SOP` both? If so, should there be `hindex` that is overloaded between the two?

2. It's also somewhat strange that there is no `hapInjs` despite there being two different functions available called `apInjs_NP` and `apInjs_POP`.

The general policy seems to be to export an overloaded function from `Generics.SOP` and make the individual functions available only when importing `Generics.SOP.NP` and `Generics.SOP.NS`.

I'm currently inclined to make everything more uniform, and add both `hindex` and `hapInjs`. I'll probably keep exporting `apInjs_NP` and `apInjs_POP` from `Generics.SOP` for backwards compatibility though.

@phadej Does this sound good?

**Comment by phadej** [comment]
I agree. `hindex` directly on `SOP` would be useful. We could make a little bigger step, and re-define `apInjs_NP` and `apINjs_POP` in `Generics.SOP` and deprecate those with a message: "Use hapInjs or import appropriate module to get warning free version". But I don't insist on that.

**Comment by kosmikus** [comment]
@phadej The deprecation trick you suggest seems to have the problem that if a user imports
```
import Generics.SOP
import Generics.SOP.NS
```
(as I guess is still reasonably common) and then calls `apInjs_NS`, then they'll suddenly get an error now because the identifier points to two different versions of the function, one being deprecated, but not the other. Getting an error from one version to the next rather than a warning seems like a worse solution overall. 

There seems to be GHC Trac issue 4879 for deprecating exports itself, but it doesn't look like it's implemented.

**Comment by phadej** [comment]
True, didn't think about it.

Maybe I'll try to hack [GHC-4879](https://ghc.haskell.org/trac/ghc/ticket/4879) next time I want to improve something in GHC :)

Your plan is :+1:

**Comment by kosmikus** [comment]
Closed via #24.
